proposals/0015-either.md

```diff
@@ -0,0 +1,207 @@
+		case let .Right(r):
+			return .Right(r)
+		}
+	}
+
+	/// If the `Either` is `Right`, simply returns a new `Right` with
+	/// the value of the receiver. If `Left`, applies the function `f`
+	/// and returns the result.
+	public func flatMap<U>(@noescape f: (LeftValue) throws -> Either<U, RightValue>) rethrows -> Either<U, RightValue> {
+		switch self {
+		case let .Left(l):
+			return try f(l)
+		case let .Right(r):
+			return .Right(r)
+		}
+	}
+
+	/// Case analysis for the `Either` type.
+	///
+	/// If the value is `.Left(a)`, apply the first function to `a`. If it is
+	/// `.Right(b)`, apply the second function to `b`.
+	public func either<U>(@noescape onLeft onLeft : (LeftValue) throws -> U, @noescape onRight : (RightValue) throws -> U) rethrows -> U {
+		switch self {
+		case let .Left(e):
+			return try onLeft(e)
+		case let .Right(e):
+			return try onRight(e)
+		}
+	}
+
+	/// Reverses the order of values of the receiver.
+	public var flip : Either<RightValue, LeftValue> {
+		switch self {
+		case let .Left(l):
+			return .Right(l)
+		case let .Right(r):
+			return .Left(r)
+		}
+	}
+
+	/// Determines if this `Either` value is a `Left`.
+	public var isLeft : Bool {
+		switch self {
+		case .Left(_):
+			return true
+		case .Right(_):
+			return false
+		}
+	}
+
+	/// Determines if this `Either` value is a `Right`.
+	public var isRight : Bool {
+		switch self {
+		case .Right(_):
+			return true
+		case .Left(_):
+			return false
+		}
+	}
+}
+
+extension Either : CustomStringConvertible {
+	/// A textual representation of `self`.
+	public var description : String {
+		switch self {
+		case let .Left(l):
+			return "Left(\(l))"
+		case let .Right(r):
+			return "Right(\(r))"
+		}
+	}
+}
+
+public func == <LeftValue : Equatable, RightValue : Equatable>(lhs : Either<LeftValue, RightValue>, rhs : Either<LeftValue, RightValue>) -> Bool {
+	switch (lhs, rhs) {
+	case let (.Left(l), .Left(r)) where l == r:
+		return true
+	case let (.Right(l), .Right(r)) where l == r:
+		return true
+	default:
+		return false
+	}
+}
+
+public func != <LeftValue : Equatable, RightValue : Equatable>(lhs : Either<LeftValue, RightValue>, rhs : Either<LeftValue, RightValue>) -> Bool {
+	return !(lhs == rhs)
+}
+```
+
+## Impact on existing code
+
+As this is an addition to the Standard Library, no existing code should be affected
+unless the author happens to be using the identifier Either - in which case
+there is a strong chance that their Either and our Either are the same
+thing, aligning with the goal of removing duplicate implementations of
+this common type.
+
+The fears of the previous proposal that attempted to put a `Result<T>`
+type in the Standard Library, of duplicating existing functionality with respect to
+`throws` remain, but as noted before `throws` is not as
+flexible or declarative enough for all possible cases.  If necessary,
+a note can be left in the documentation warning away users of the
+`Either` type that really need `throws`.
```

Comment:
Further improvement of https://github.com/TypeLift/Focus is effectively blocked on the availability of a universal `Either`, since it's crucial to more powerful prisms, as well as even the mere existence of traversals and partial lenses.